fdbclient/MultiVersionTransaction.actor.cpp

```diff
@@ -871,12 +877,92 @@ ThreadFuture<Void> MultiVersionTransaction::onError(Error const& e) {
 	}
 }
 
+// Waits for the specified duration and signals the assignment variable with a timed out error
+// This will be canceled if a new timeout is set, in which case the tsav will not be signaled.
+ACTOR Future<Void> timeoutImpl(Reference<ThreadSingleAssignmentVar<Void>> tsav, double duration) {
+	wait(delay(duration));
+
+	tsav->trySendError(transaction_timed_out());
+	return Void();
+}
+
+// Configure a timeout based on the options set for this transaction. This timeout only applies
+// if we don't have an underlying database object to connect with.
+void MultiVersionTransaction::setTimeout(Optional<StringRef> value) {
+	double timeoutDuration = extractIntOption(value, 0, std::numeric_limits<int>::max()) / 1000.0;
+
+	ThreadFuture<Void> prevTimeout;
+	ThreadFuture<Void> newTimeout = onMainThread([this, timeoutDuration]() {
+		return timeoutImpl(timeoutTsav, timeoutDuration - std::max(0.0, now() - startTime));
+	});
+
+	{ // lock scope
+		ThreadSpinLockHolder holder(timeoutLock);
+
+		prevTimeout = currentTimeout;
+		currentTimeout = newTimeout;
+	}
+
+	// Cancel the previous timeout now that we have a new one. This means that changing the timeout
+	// affects in-flight operations, which is consistent with the behavior in RYW.
+	if (prevTimeout.isValid()) {
+		prevTimeout.cancel();
```

Comment:
Is this necessary? I thought this would be cancelled anyways when `prevTimeout` is running out of scope?

Reply:
Yeah, I think you're right. I believe a previous version shared the SAV here and so we needed explicit cancelation, but that shouldn't be required now.

Reply:
For whatever reason, removing this did not have the same outcome. For now I'll leave it in and we can resolve the question of why later.